src/gui/symbology/qgscategorizedsymbolrendererwidget.cpp

```diff
@@ -1093,6 +1067,62 @@ void QgsCategorizedSymbolRendererWidget::deleteAllCategories()
   emit widgetChanged();
 }
 
+void QgsCategorizedSymbolRendererWidget::deleteUnusedCategories()
+{
+  if ( !mRenderer )
+    return;
+  QString attrName = mExpressionWidget->currentField();
+  QList<QVariant> uniqueValues = layerUniqueValues( attrName );
+
+  const QgsCategoryList catList = mRenderer->categories();
+
+  QList<int> unusedIndexes;
+
+  for ( int i = 0; i < catList.size(); ++i )
+  {
+    QgsRendererCategory cat = catList.at( i );
+    if ( !uniqueValues.contains( cat.value() ) )
+    {
+      unusedIndexes.append( i );
+    }
+  }
+  mModel->deleteRows( unusedIndexes );
+  emit widgetChanged();
+}
+
+QList<QVariant> QgsCategorizedSymbolRendererWidget::layerUniqueValues( QString attrName )
```

Comment:
```suggestion
QList<QVariant> QgsCategorizedSymbolRendererWidget::layerUniqueValues( const QString& attrName )
```

Even if QString is implicit data shared, all the code use const ref for string, so let's do the same out of consistency